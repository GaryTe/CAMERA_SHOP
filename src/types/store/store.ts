import { store } from '../../store/store/store';
import { listCamera, CardProduct, Reviews, PromoProduct, Review } from '../types-response/types-response';

export type Cameras = {
  cameras: listCamera | null;
  modifiedListProduct: listCamera | null;
}

export type CardCamera = {
  dataCamera: CardProduct | null;
}

export type SimilaCameras = {
  dataSimilaCameras: listCamera | null;
  sortListCamerra: listCamera | null;
  counterMin: number;
  counterMax: number;
}

export type ReviewsCamera = {
  dataReviews: Reviews | Array<Review>;
  resultSortReviews: Reviews | null;
  reviews: Reviews | null;
  counter: number;
}

export type PromoCamera = {
  dataPromoCamera: PromoProduct | null;
}

export type Cart = {
  dataProduct: listCamera | Array<CardProduct>;
  loading: boolean;
}

export type RootState = ReturnType<typeof store.getState>;
