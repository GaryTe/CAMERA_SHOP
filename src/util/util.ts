import {MutableRefObject, KeyboardEvent} from 'react';
import { Reviews, listCamera } from '../types/types-response/types-response';
import { Category, TypeCamera, LevelProduct } from '../const/const';

import dayjs from 'dayjs';


export const gettingTranslatedDate = (date: string) => dayjs (date).format ('D MMMM');


export const sortReviews = (reviews: Reviews) => {
  let resultSort: Reviews = [];

  if(reviews.length === 1) {
    resultSort = reviews;
  } else {
    resultSort = reviews.slice ().sort ((a,b) => {
      if (a.createAt < b.createAt) {
        return -1;
      }
      if (a.createAt > b.createAt) {
        return 1;
      }
      return 0;
    }).reverse();
  }

  return resultSort;
};


export const down = (products: listCamera, payload: string) => {
  let resultSort: listCamera = [];

  resultSort = products.slice ().sort ((a,b) => {
    if (a[payload] < b[payload]) {
      return -1;
    }
    if (a[payload] > b[payload]) {
      return 1;
    }
    return 0;
  }).reverse();

  return resultSort;
};

export const up = (products: listCamera, payload: string) => {
  let resultSort: listCamera = [];

  resultSort = products.slice ().sort ((a,b) => {
    if (a[payload] < b[payload]) {
      return -1;
    }
    if (a[payload] > b[payload]) {
      return 1;
    }
    return 0;
  });

  return resultSort;
};


export const filteringByCategory = (products: listCamera, payload: string): listCamera => {
  let filteredListProduct: listCamera = [];

  switch (payload) {
    case Category.Camera:
      filteredListProduct = products.slice().filter((camera) => camera.category === Category.Camera);
      break;
    case Category.Videocamera:
      filteredListProduct = products.slice().filter((Videocamera) => Videocamera.category === Category.Videocamera);
      break;
    case TypeCamera.Digital:
      filteredListProduct = products.slice().filter((camera) => camera.type === TypeCamera.Digital);
      break;
    case TypeCamera.Instant:
      filteredListProduct = products.slice().filter((camera) => camera.type === TypeCamera.Instant);
      break;
    case TypeCamera.Membranous:
      filteredListProduct = products.slice().filter((camera) => camera.type === TypeCamera.Membranous);
      break;
    case TypeCamera.Collection:
      filteredListProduct = products.slice().filter((camera) => camera.type === TypeCamera.Collection);
      break;
    case LevelProduct.Zero:
      filteredListProduct = products.slice().filter((camera) => camera.level === LevelProduct.Zero);
      break;
    case LevelProduct.Professional:
      filteredListProduct = products.slice().filter((camera) => camera.level === LevelProduct.Professional);
      break;
    case LevelProduct.Amateur:
      filteredListProduct = products.slice().filter((camera) => camera.level === LevelProduct.Amateur);
      break;
  }

  return filteredListProduct;
};

export const filteringByPrice = (
  products: listCamera,
  priceFrom: number | null,
  priceTo: number | null
) => {
  let totalProducts: listCamera = [];

  if(priceFrom && !priceTo) {
    totalProducts = products.filter((product) => product.price > priceFrom);
  }
  if(!priceFrom && priceTo) {
    totalProducts = products.filter((product) => product.price < priceTo);
  }
  if(priceFrom && priceTo) {
    totalProducts = products.filter((product) => product.price > priceFrom && product.price < priceTo);
  }

  return totalProducts;
};


export const returnListNumber = (
  value: number
) =>{
  let counter = 0;
  const items = [];

  while(counter <= value) {
    counter += 1;

    items.push(counter);
  }
  return items;
};


export const returnNumber = (number: number) => {
  const valueNumber = (number / 100) * 100;
  let result = 0;

  if(valueNumber > 0 && valueNumber <= 20) {
    result = 1;
  }
  if(valueNumber > 20 && valueNumber <= 40) {
    result = 2;
  }
  if(valueNumber > 40 && valueNumber <= 60) {
    result = 3;
  }
  if(valueNumber > 60 && valueNumber <= 80) {
    result = 4;
  }
  if(valueNumber > 80 && valueNumber > 100) {
    result = 5;
  }

  return result;
};

export const shiftFocus = (evt: KeyboardEvent<HTMLDivElement>, element: MutableRefObject<null | HTMLElement>) => {
  if(evt.key === 'Tab') {
    element.current?.focus();

  }
};
