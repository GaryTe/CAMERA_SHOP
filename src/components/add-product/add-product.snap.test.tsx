import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import AddProduct from './add-product';
import { product } from '../../util/mocks';
import { store } from '../../store/store/store';

describe('Component: AddProduct', () => {
  it('should render AddProduct', () => {

    const {container} = render(
      <Provider store={store}>
        <AddProduct
          dataProduct={product}
          onSetModalWindow={jest.fn()}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
