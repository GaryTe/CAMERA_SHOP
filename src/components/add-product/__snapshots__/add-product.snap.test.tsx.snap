// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AddProduct should render AddProduct 1`] = `
<div>
  <div
    class="modal is-active"
  >
    <div
      class="modal__wrapper"
    >
      <div
        class="modal__overlay"
      />
      <div
        class="modal__content"
      >
        <p
          class="title title--h4"
        >
          Добавить товар в корзину
        </p>
        <div
          class="basket-item basket-item--short"
        >
          <div
            class="basket-item__img"
          >
            <picture>
              <source
                src="img/content/das-auge.webp"
                srcset="img/content/das-auge@2x.webp"
                type="image/webp"
              />
              <img
                alt="Фотоаппарат «Орлёнок»"
                height="120"
                src="img/content/das-auge.jpg"
                srcset="img/content/das-auge@2x.jpg"
                width="140"
              />
            </picture>
          </div>
          <div
            class="basket-item__description"
          >
            <p
              class="basket-item__title"
            >
              Ретрокамера Dus Auge lV
            </p>
            <ul
              class="basket-item__list"
            >
              <li
                class="basket-item__list-item"
              >
                <span
                  class="basket-item__article"
                >
                  Артикул:
                </span>
                <span
                  class="basket-item__number"
                >
                  DA4IU67AD5
                </span>
              </li>
              <li
                class="basket-item__list-item"
              >
                Коллекционная
                 фотокамера
              </li>
              <li
                class="basket-item__list-item"
              >
                Нулевой
                 уровень
              </li>
            </ul>
            <p
              class="basket-item__price"
            >
              <span
                class="visually-hidden"
              >
                Цена:
              </span>
              65000
            </p>
          </div>
        </div>
        <div
          class="modal__buttons"
        >
          <button
            class="btn btn--purple modal__btn modal__btn--fit-width"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="16"
              width="24"
            >
              <use
                xlink:href="#icon-add-basket"
              />
            </svg>
            Добавить в корзину
          </button>
        </div>
        <button
          aria-label="Закрыть попап"
          class="cross-btn"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="10"
            width="10"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
